File: backend-code/app/api/routes/agents.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List
from app.core.database import get_db
from app.utils.dependencies import get_current_active_user
from app.models.user import User as UserModel
from app.schemas.agent import Agent, AgentCreate, AgentUpdate
from app.models.agent import agent as agent_crud

router = APIRouter()


@router.get("/", response_model=List[Agent])
def get_agents(
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db),
    current_user: UserModel = Depends(get_current_active_user)
):
    """获取当前用户的Agent列表"""
    agents = agent_crud.get_multi(db, user_id=current_user.id, skip=skip, limit=limit)
    return agents


@router.get("/default", response_model=Agent)
def get_default_agent(
    db: Session = Depends(get_db),
    current_user: UserModel = Depends(get_current_active_user)
):
    """获取当前用户的默认Agent"""
    default_agent = agent_crud.get_default_by_user(db, user_id=current_user.id)
    if not default_agent:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="No default agent found"
        )
    return default_agent


@router.post("/", response_model=Agent)
def create_agent(
    agent_in: AgentCreate,
    db: Session = Depends(get_db),
    current_user: UserModel = Depends(get_current_active_user)
):
    """创建新的Agent"""
    # 如果是第一个Agent，自动设为默认
    existing_agents = agent_crud.get_by_user(db, user_id=current_user.id)
    if not existing_agents:
        agent_in.is_default = True
    
    agent_data = agent_in.dict()
    agent_data["user_id"] = current_user.id
    
    agent = agent_crud.create(db, obj_in=agent_data)
    return agent


@router.put("/{agent_id}", response_model=Agent)
def update_agent(
    agent_id: int,
    agent_in: AgentUpdate,
    db: Session = Depends(get_db),
    current_user: UserModel = Depends(get_current_active_user)
):
    """更新Agent"""
    agent = agent_crud.get(db, id=agent_id)
    if not agent:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Agent not found"
        )
    
    if agent.user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not authorized to update this agent"
        )
    
    update_data = agent_in.dict(exclude_unset=True)
    agent = agent_crud.update(db, db_obj=agent, obj_in=update_data)
    return agent


@router.delete("/{agent_id}")
def delete_agent(
    agent_id: int,
    db: Session = Depends(get_db),
    current_user: UserModel = Depends(get_current_active_user)
):
    """删除Agent"""
    agent = agent_crud.get(db, id=agent_id)
    if not agent:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Agent not found"
        )
    
    if agent.user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not authorized to delete this agent"
        )
    
    # 如果删除的是默认Agent，需要将其他Agent设为默认
    if agent.is_default:
        other_agents = agent_crud.get_by_user(db, user_id=current_user.id)
        other_agents = [a for a in other_agents if a.id != agent_id]
        if other_agents:
            agent_crud.set_default(db, agent_id=other_agents[0].id, user_id=current_user.id)
    
    agent_crud.delete(db, id=agent_id)
    return {"message": "Agent deleted successfully"}


@router.put("/{agent_id}/set-default", response_model=Agent)
def set_default_agent(
    agent_id: int,
    db: Session = Depends(get_db),
    current_user: UserModel = Depends(get_current_active_user)
):
    """设置默认Agent"""
    agent = agent_crud.get(db, id=agent_id)
    if not agent:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Agent not found"
        )
    
    if agent.user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not authorized to update this agent"
        )
    
    agent = agent_crud.set_default(db, agent_id=agent_id, user_id=current_user.id)
    return agent


@router.get("/{agent_id}", response_model=Agent)
def get_agent(
    agent_id: int,
    db: Session = Depends(get_db),
    current_user: UserModel = Depends(get_current_active_user)
):
    """获取特定Agent"""
    agent = agent_crud.get(db, id=agent_id)
    if not agent:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Agent not found"
        )
    
    if agent.user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not authorized to access this agent"
        )
    
    return agent

File: frontend-code-generation/lib/services/agents.ts
import { api } from '@/lib/api'

export interface Agent {
  id: number
  user_id: number
  name: string
  description?: string
  prompt: string
  icon: string
  is_active: boolean
  is_default: boolean
  created_at: string
  updated_at?: string
}

export interface AgentCreate {
  name: string
  description?: string
  prompt: string
  icon?: string
  is_active?: boolean
  is_default?: boolean
}

export interface AgentUpdate {
  name?: string
  description?: string
  prompt?: string
  icon?: string
  is_active?: boolean
  is_default?: boolean
}

// 使用API客户端的ApiResponse接口，避免类型冲突
export interface AgentServiceResponse<T> {
  data?: T
  error?: string
  message?: string
  status: 'success' | 'error'
}

export const agentsService = {
  // 获取用户的Agent列表
  async getAgents(): Promise<AgentServiceResponse<Agent[]>> {
    try {
      console.log('🔍 [Agent服务] 开始获取Agent列表')
      const response = await api.get('/api/agents')
      console.log('📥 [Agent服务] 获取Agent列表响应:', response)
      
      return {
        data: response.data as Agent[],
        status: 'success'
      }
    } catch (error: any) {
      console.error('❌ [Agent服务] 获取Agent列表失败:', error)
      console.error('   🔍 错误详情:', {
        message: error.message,
        response: error.response?.data,
        status: error.response?.status
      })
      
      return {
        error: error.response?.data?.detail || error.message || '获取Agent列表失败',
        status: 'error'
      }
    }
  },

  // 获取默认Agent
  async getDefaultAgent(): Promise<AgentServiceResponse<Agent>> {
    try {
      console.log('🔍 [Agent服务] 开始获取默认Agent')
      const response = await api.get('/api/agents/default')
      console.log('📥 [Agent服务] 获取默认Agent响应:', response)
      
      return {
        data: response.data as Agent,
        status: 'success'
      }
    } catch (error: any) {
      console.error('❌ [Agent服务] 获取默认Agent失败:', error)
      console.error('   🔍 错误详情:', {
        message: error.message,
        response: error.response?.data,
        status: error.response?.status
      })
      
      return {
        error: error.response?.data?.detail || error.message || '获取默认Agent失败',
        status: 'error'
      }
    }
  },

  // 创建新Agent
  async createAgent(agentData: AgentCreate): Promise<AgentServiceResponse<Agent>> {
    try {
      console.log('🔍 [Agent服务] 开始创建Agent:', agentData)
      const response = await api.post('/api/agents', agentData)
      console.log('📥 [Agent服务] 创建Agent响应:', response)
      console.log('📦 [Agent服务] 响应数据:', response.data)
      
      return {
        data: response.data as Agent,
        status: 'success'
      }
    } catch (error: any) {
      console.error('❌ [Agent服务] 创建Agent失败:', error)
      console.error('   🔍 错误详情:', {
        message: error.message,
        response: error.response?.data,
        status: error.response?.status,
        statusText: error.response?.statusText
      })
      
      return {
        error: error.response?.data?.detail || error.message || '创建Agent失败',
        status: 'error'
      }
    }
  },

  // 更新Agent
  async updateAgent(id: number, agentData: AgentUpdate): Promise<AgentServiceResponse<Agent>> {
    try {
      console.log('🔍 [Agent服务] 开始更新Agent:', id, agentData)
      const response = await api.put(`/api/agents/${id}`, agentData)
      console.log('📥 [Agent服务] 更新Agent响应:', response)
      
      return {
        data: response.data as Agent,
        status: 'success'
      }
    } catch (error: any) {
      console.error('❌ [Agent服务] 更新Agent失败:', error)
      console.error('   🔍 错误详情:', {
        message: error.message,
        response: error.response?.data,
        status: error.response?.status
      })
      
      return {
        error: error.response?.data?.detail || error.message || '更新Agent失败',
        status: 'error'
      }
    }
  },

  // 删除Agent
  async deleteAgent(id: number): Promise<AgentServiceResponse<void>> {
    try {
      console.log('🔍 [Agent服务] 开始删除Agent:', id)
      await api.delete(`/api/agents/${id}`)
      console.log('✅ [Agent服务] 删除Agent成功')
      
      return {
        status: 'success'
      }
    } catch (error: any) {
      console.error('❌ [Agent服务] 删除Agent失败:', error)
      console.error('   🔍 错误详情:', {
        message: error.message,
        response: error.response?.data,
        status: error.response?.status
      })
      
      return {
        error: error.response?.data?.detail || error.message || '删除Agent失败',
        status: 'error'
      }
    }
  },

  // 设置默认Agent
  async setDefaultAgent(id: number): Promise<AgentServiceResponse<Agent>> {
    try {
      console.log('🔍 [Agent服务] 开始设置默认Agent:', id)
      const response = await api.put(`/api/agents/${id}/set-default`)
      console.log('📥 [Agent服务] 设置默认Agent响应:', response)
      
      return {
        data: response.data as Agent,
        status: 'success'
      }
    } catch (error: any) {
      console.error('❌ [Agent服务] 设置默认Agent失败:', error)
      console.error('   🔍 错误详情:', {
        message: error.message,
        response: error.response?.data,
        status: error.response?.status
      })
      
      return {
        error: error.response?.data?.detail || error.message || '设置默认Agent失败',
        status: 'error'
      }
    }
  },

  // 获取特定Agent
  async getAgent(id: number): Promise<AgentServiceResponse<Agent>> {
    try {
      const response = await api.get(`/api/agents/${id}`)
      return {
        data: response.data as Agent,
        status: 'success'
      }
    } catch (error: any) {
      console.error('获取Agent失败:', error)
      return {
        error: error.response?.data?.detail || '获取Agent失败',
        status: 'error'
      }
    }
  }
}